### ISSUE TITLE
Incorrect `error: Unnecessary isinstance call` related to dict

### ISSUE THREAD

**Issue by EqualMa** [microsoft/pyright#1149]
**Describe the bug**
`Unnecessary isinstance call` error is reported when writing something like:

```python
value: Union[SomeTypedDict, int]
isinstance(value, dict) # error: Unnecessary isinstance call
```

It seems that pyright doesn't consider `SomeTypedDict` as sub-type of `Dict[Unkown, Unkown]`

**To Reproduce**

Use pyright to check the code in *Screenshots or Code*.

And pyright will report the following errors:

```
  10:8 - error: Unnecessary isinstance call; "Person | int" is never an instance of "dict[Unknown, Unknown]" (reportUnnecessaryIsInstance)
  13:36 - error: Argument of type "Person | int" cannot be assigned to parameter "age" of type "int" in function "__init__"
    Type "Person | int" cannot be assigned to type "int"
      "Person" is incompatible with "int" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 infos 
```

**Expected behavior**

No errors should be reported.

**Screenshots or Code**

```python
from typing import TypedDict, Union


class Person(TypedDict):
    name: str
    age: int


def parse_person(person: Union[Person, int]) -> Person:
    if isinstance(person, dict):
        return Person(**person)
    else:
        return Person(name="", age=person)
```

**VS Code extension or command-line**

pyright 1.1.79

**Additional context**

About why I use `dict` instead of `SomeTypedDict` in `isinstance`:

> In particular, TypedDict type objects cannot be used in isinstance() tests such as isinstance(d, Movie).
> [pep-0589/#using-typeddict-types](https://www.python.org/dev/peps/pep-0589/#using-typeddict-types)

**Comment by erictraut** [comment]
Thanks for the bug report. You're correct that TypedDict objects require special-case logic here. From a type-checking perspective, a TypedDict is not compatible with a dict (i.e. you cannot assign a TypedDict to a variable that has a declared type of `dict`), but at runtime the `isinstance` call will return True when you pass `dict` as the second argument.

I've updated the "unnecessary isinstance call" check as well as the type narrowing logic so they know about this special case.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.84, which I just published. It will also be included in the next release of Pylance.

**Comment by EqualMa** [comment]
Thanks for your work!
I just realized that TypedDict is not sub-type of Dict as explained in [pep-0589](https://www.python.org/dev/peps/pep-0589/#type-consistency):
> A TypedDict isn't consistent with any Dict[...] type, since dictionary types allow destructive operations, including clear(). They also allow arbitrary keys to be set, which would compromise type safety.
